finagle-base-http/src/main/scala/com/twitter/finagle/http/headers/JTreeMapBackedHeaderMap.scala

```diff
@@ -0,0 +1,79 @@
+package com.twitter.finagle.http.headers
+
+import com.twitter.finagle.http.HeaderMap
+import scala.annotation.tailrec
+
+/**
+ * Mutable, thread-safe [[HeaderMap]] implementation, backed by
+ * a mutable [[Map[String, Header]]], where the map key
+ * is forced to lower case
+ */
+final private[http] class JTreeMapBackedHeaderMap extends JMapBackedHeaderMap {
+  
+  override val underlying: java.util.TreeMap[String, Header.Root] = 
+    new java.util.TreeMap[String, Header.Root](JTreeMapBackedHeaderMap.SharedComparitor)
+
+    def getAll(key: String): Seq[String] = underlying.synchronized {
+      underlying.get(key.toLowerCase) match {
+        case null => Nil
+        case r: Header.Root => r.values
+      }
+    }
+
+  // Does not validate key and value.
+  def addUnsafe(key: String, value: String): this.type  = underlying.synchronized {
+    def header = Header.root(key, value)
```

Comment:
`header` is only needed if we hit the `null` case. Can we move its instantiation there?

Reply:
It's a def for that reason, but it can go inside the match. The only reason it's outside now is neater indentation, so that's not much of a reason.

Reply:
Oh, I didn't notice it was a def. :) Maybe just inline it where `header` is used and we'll both be happy.

As an aside, do you know if that def allocates a lambda, or does that get lifted out into the object itself?

Reply:
Yes, I inlined it.

As for the allocation, I assumed it gets lifted to a method all the way to the top, especially since it doesn't capture anything and doesn't escape scope, but I'm not 100% sure either way.